client/go/dogma/watch_service.go

```diff
@@ -0,0 +1,358 @@
+// under the License.
+
+package dogma
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+	"unsafe"
+)
+
+type watchService service
+
+// WatchResult represents a result from watch operation.
+type WatchResult struct {
+	Commit *Commit
+	Res    *http.Response
+	Err    error
+}
+
+func (ws *watchService) watchFile(ctx context.Context, projectName, repoName, lastKnownRevision string,
+	query *Query, timeout time.Duration) <-chan *WatchResult {
+	watchResult := make(chan *WatchResult)
+	if query == nil {
+		watchResult <- &WatchResult{Err: errors.New("query should not be nil")}
+		return watchResult
+	}
+
+	u := fmt.Sprintf("%vprojects/%v/repos/%v/contents%v", defaultPathPrefix, projectName, repoName, query.Path)
+	v := &url.Values{}
+	if query != nil && query.Type == JSONPath {
+		if err := setJSONPaths(v, query.Path, query.Expressions); err != nil {
+			watchResult <- &WatchResult{Err: err}
+			return watchResult
+		}
+	}
+	u += encodeValues(v)
+	ws.watchRequest(ctx, watchResult, u, lastKnownRevision, timeout)
+	return watchResult
+}
+
+func (ws *watchService) watchRepo(ctx context.Context,
+	projectName, repoName, lastKnownRevision, pathPattern string, timeout time.Duration) <-chan *WatchResult {
+	watchResult := make(chan *WatchResult)
+	if len(pathPattern) != 0 && !strings.HasPrefix(pathPattern, "/") {
+		// Normalize the pathPattern when it does not start with "/" so that the pathPattern fits into the url.
+		pathPattern = "/**/" + pathPattern
+	}
+
+	u := fmt.Sprintf("%vprojects/%v/repos/%v/contents%v", defaultPathPrefix, projectName, repoName, pathPattern)
+	ws.watchRequest(ctx, watchResult, u, lastKnownRevision, timeout)
+	return watchResult
+}
+
+func (ws *watchService) watchRequest(ctx context.Context, watchResult chan<- *WatchResult,
+	u, lastKnownRevision string, timeout time.Duration) {
+	req, err := ws.client.newRequest("GET", u, nil)
+	if err != nil {
+		watchResult <- &WatchResult{Err: err}
+		return
+	}
+	if len(lastKnownRevision) != 0 {
+		req.Header.Set("if-none-match", lastKnownRevision)
+	} else {
+		req.Header.Set("if-none-match", "-1")
+	}
+	if timeout != 0 {
+		req.Header.Set("prefer", fmt.Sprintf("wait=%v", timeout.Seconds()))
+	}
+
+	go func() {
+		commit := new(Commit)
+		res, err := ws.client.do(ctx, req, commit)
+		if err != nil {
+			watchResult <- &WatchResult{Commit: nil, Res: res, Err: err}
+		} else {
+			watchResult <- &WatchResult{Commit: commit, Res: res, Err: nil}
+		}
+	}()
+}
+
+const watchTimeout = 1 * time.Minute
+
+// These constants represent the state of a watcher.
+const (
+	initial int32 = iota
+	started
+	stopped
+)
+
+// Watcher watches the changes of a repository or a file.
+type Watcher struct {
+	state              int32
+	initialChan        chan *Latest // channel whose buffer is 1.
+	isInitialChanSet   int32        // 0 is false, 1 is true
+	scheduleCTX        context.Context
```

Comment:
just question, if `scheduleCTX` and `watchCTX` mean to be called in `Close()` only, why do we need a lot of contexts? What will happen if we have single `watcherCTX` for a single watcher and handle everything (closed state, close signal dispatcher...) through that?

Reply:
You are right~! I don't have to use two different context here. Thank you~!